_posts/2020-09-25-programmatic-authentication-under-iap.md

```diff
@@ -0,0 +1,200 @@
+
+### The Solution
+
+Fortunately, Google has made google auth libraries where all the above procedures are already implemented and are well abstracted. Searching google-auth-library would return results with their implementation in multiple languages.
+We used the [node library](https://github.com/googleapis/google-auth-library-nodejs) since we had our automation projects majorly implemented through protractor and javascript.
+The implementation looks something like this :
+
+```javascript
+function  main(
+   targetAudience  =  'CLIENT_ID_GOES_HERE'
+) {
+   const { GoogleAuth } =  require('google-auth-library');
+   const  auth  =  new  GoogleAuth();
+
+   function  request() {
+       auth.getIdTokenClient(targetAudience).then(function(client){
+           client.getRequestHeaders().then(function(headers){
+               console.info("headers", headers);
+           });
+       });
+   }
+   request();
+}
+
+main();
+```
+
+This was used to generate the Bearer auth token that could then be further used in making requests to our test apps under automation testing.
+
+In the app automation where we required browser instances from Protractor to run automations on test-apps, we made use of the node implementation of [browsermob-proxy](https://github.com/zzo/browsermob-node) along with the [browsermob-proxy server](https://github.com/lightbody/browsermob-proxy).
+Using Protractor’s lifecycle method `beforeAll`, we made sure that the token is fetched and ready to be used before any of the spec files are executed.
+
+```javascript
+exports.setIap = function () {
+   beforeAll(function (done) {
+       global.iapAuthToken = 'CLIENT_ID_GOES_HERE';
+       if (browser.params.isRunOnTestApp) {
+           var targetAudience = '';
+           const auth = new GoogleAuth();
+           auth.getIdTokenClient(targetAudience).then(function (client) {
+               client.getRequestHeaders().then(function (headers) {
+                   iapAuthToken = headers.Authorization;
+                   done();
+               });
+           });
+       } else {
+           done();
+       }
+   });
+}
+
+```
+
+Then the global available value of this token was used to set the authorization header in the proxy spawned.
+We used the `addHeader` method for this.
+
+```javascript
+var headersToSet = {};
+headersToSet.Authorization = global.iapAuthToken;
+proxy.addHeader(port, headersToSet, function (err, resp) {
+   if (err) {
+       console.error("Error encountered", err);
+   } else {
+       console.info(`Headers added ${JSON.stringify(headersToSet)}, ${resp}`);
+   }
+});
+```
+
+This allowed us to add the authorization bearer token to every request that was being made through the Protractor browser instance.
+Additionally, we also had to set the `trustAllServers` parameter in browsermob-proxy to `true` to disable verification of all upstream servers' SSL certificates.
```

Comment:
this line is extra information, can be removed.

Reply:
removed